src/ucp/tag/tag_send.c

```diff
@@ -355,7 +303,74 @@ UCS_PROFILE_FUNC(ucs_status_ptr_t, ucp_tag_send_nbx,
                  ucp_ep_h ep, const void *buffer, size_t count,
                  ucp_tag_t tag, const ucp_request_param_t *param)
 {
-    return UCS_STATUS_PTR(UCS_ERR_NOT_IMPLEMENTED);
+    size_t contig_length = 0;
+    ucs_status_t status;
+    ucp_request_t *req;
+    ucs_status_ptr_t ret;
+    uintptr_t datatype;
+    uint32_t attr_mask;
+    ucp_worker_h worker;
+    ucp_send_callback_t cb;
+
+    UCP_CONTEXT_CHECK_FEATURE_FLAGS(ep->worker->context, UCP_FEATURE_TAG,
+                                    return UCS_STATUS_PTR(UCS_ERR_INVALID_PARAM));
+    UCP_WORKER_THREAD_CS_ENTER_CONDITIONAL(ep->worker);
+
+    ucs_trace_req("send_nbx buffer %p count %zu tag %" PRIx64 " to %s",
+                  buffer, count, tag, ucp_ep_peer_name(ep));
+
+    attr_mask = param->op_attr_mask &
+                (UCP_OP_ATTR_FIELD_DATATYPE | UCP_OP_ATTR_FLAG_NO_IMM_CMPL);
+
+    if (ucs_likely(attr_mask == 0)) {
+        status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer, count, 
+                                  ucp_dt_make_contig(1), tag);
+        ucp_request_send_check_status(status, ret, goto out);
+        datatype = ucp_dt_make_contig(1);
+        contig_length = count;
+    } else if (attr_mask == UCP_OP_ATTR_FIELD_DATATYPE) {
+        datatype = param->datatype;
+        if (ucs_likely(UCP_DT_IS_CONTIG(datatype))) {
+            contig_length = ucp_contig_dt_length(datatype, count);
+            status = UCS_PROFILE_CALL(ucp_tag_send_inline, ep, buffer,
+                                      contig_length, datatype, tag);
+            ucp_request_send_check_status(status, ret, goto out);
+        }
+    } else {
+        datatype = ucp_dt_make_contig(1);
+    }
+
+    if (ucs_unlikely(param->op_attr_mask & UCP_OP_ATTR_FLAG_FORCE_IMM_CMPL)) {
+        ret = UCS_STATUS_PTR(UCS_ERR_NO_RESOURCE);
+        goto out;
+    }
+
+    worker = ep->worker;
+    req = ucp_request_get_param(worker, param,
+                                {
+                                    ret = UCS_STATUS_PTR(UCS_ERR_NO_MEMORY);
+                                    goto out;
+                                });
+
+    if (param->op_attr_mask & UCP_OP_ATTR_FIELD_CALLBACK) {
+        cb             = (ucp_send_callback_t)param->cb.send;
```

Comment:
shouldn't it be nbx-type callback?

Reply:
> shouldn't it be nbx-type callback?

yes... but there is another issue: there is no tag_recv_nbx call where user_data may be used - no unification.
refactored